### PR TITLE
Support for different readers

### LINKED ISSUE #4: Use native pandas read_csv

### PR DESCRIPTION
**Pull request by scravy**
This adds a new argument to `s3.select` which allows to specify a different reader. This allows for example reading JSON instead of a Pandas Data Frame, or directly getting the data into a python list.

Further work: Using this pandas could be made an optional dependency.

This fixes #4

### CODE REVIEW

File: s3access/reader.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        return NotImplemented
+
+
+class Pandas(Reader[pd.DataFrame]):
+    def read(self, bs: Union[bytes, bytearray], columns: Dict[str, Type]) -> pd.DataFrame:
+        return pd.read_csv(BytesIO(bs), header=None, names=columns.keys(), dtype=columns)
+
+    def combine(self, results: Sequence[pd.DataFrame]) -> pd.DataFrame:
+        if not results:
+            return pd.DataFrame([])
```
Comment (infradust) [inline s3access/reader.py:43]: can never decide for this case if its better to return an empty dataframe (no columns and types) or simply None.

File: s3access/s3access.py [hunk 2]
```diff
@@ -48,10 +46,13 @@ def read_value(value, type):
...
 
 
+T = TypeVar('T')
+
+
 class S3Access:
-    def __init__(self, parallelism: int = multiprocessing.cpu_count() * 4):
-        self._num_workers = parallelism
-        self._cachedir = os.getenv('S3ACCESSCACHE')
+    def __init__(self, parallelism: int = multiprocessing.cpu_count() * 4, cachedir: str = os.getenv('S3ACCESSCACHE')):
```
Comment (infradust) [inline s3access/s3access.py:53]: setting the 'cachedir' like that will create problems if importing before setting the environment variable (good for services bad for notebooks)

File: s3access/s3access.py [hunk 3]
```diff
@@ -177,21 +178,21 @@ def _check_path(path: S3Path, conditions: Dict[str, Condition]) -> bool:
     def _select_glob(self,
                      s3path: Union[str, S3Path],
                      columns: Dict[str, Type],
-                     filters: Dict[str, Condition]) -> pd.DataFrame:
+                     filters: Dict[str, Condition],
+                     reader: Reader[T]) -> T:
```
Comment (infradust) [inline s3access/s3access.py:182]: default to Pandas?

**Review by infradust — APPROVED** [review]
some questions and considerations but looks like a nice addition

File: s3access/s3access.py (hunk 3 above)
Comment (scravy) [reply s3access/s3access.py:188]: Internal method, will always get the reader passed in.

File: s3access/reader.py (hunk 1 above)
Comment (scravy) [reply s3access/reader.py:43]: The eternal question. Then the result should be `Optional[T]`.

I think an empty data frame is alright here, just as other implementations give an empty iterator, etc.

**Comment by scravy** [comment]
Fixed the reading of environment variable default value.

File: s3access/s3access.py (hunk 2 above)
Comment (scravy) [reply s3access/s3access.py]: fixed

**Pull request merged**